### PR TITLE
feat: change define.lp and add constraints to hard.lp for facets

### PR DESCRIPTION
**Pull request by Zehua-Zeng**

### CODE REVIEW

File: draco/asp/define.lp [hunk 1]
```diff
@@ -47,3 +47,12 @@ domain((encoding,stack),(zero;center;normalize)).
 
 % @definition(scale_zero) Scale start with 0.
 domain((scale,zero),true).
+
+% @definition(facets) Definition for facets
+domain((facet,field),F) :- entity(field,_,F).
+
+domain(facet_channel,(row;col)).
+domain((facet,channel),C) :- domain(facet_channel,C).
```
Comment (JunranY) [inline draco/asp/define.lp:55]:
```suggestion
domain((facet,channel),(row;col)).
```

File: draco/asp/hard.lp [hunk 2]
```diff
@@ -230,6 +230,27 @@ violation(bar_area_without_zero) :-
     not attribute((scale,zero),S,_),
     C = (x;y).
 
+% @hard(row_or_col_c) Row and column require discrete. If number, need to be binned.
+violation(row_or_col_c) :-
```
Comment (JunranY) [inline draco/asp/hard.lp:234]: I think field with number type can be discrete too. For example if it's ordinal with a small cardinality.

File: draco/asp/define.lp (hunk 1 above)
Comment (Zehua-Zeng) [reply draco/asp/define.lp]: done

File: draco/asp/hard.lp (hunk 2 above)
Comment (JunranY) [reply draco/asp/hard.lp:234]: @domoritz Do we want to write this one in soft constraints considering numeric field can be ordinal(but can't be explicitly indicated)?

Comment (domoritz) [reply draco/asp/hard.lp:234]: Yes, a numeric field can be discrete as well. We could have soft constraints that require high cardinality numbers to be binned to be discrete.

Comment (Zehua-Zeng) [reply draco/asp/hard.lp:234]: So we don't need this hard constraint to require numbers must be binned for faceting?

Comment (domoritz) [reply draco/asp/hard.lp:234]: I don't think we should have this as a hard constraint. The soft constraint should be a bit more thoughtful like "the cardinality of facets should be less than x" and then we should define the effective cardinality based on the data cardinality and binning. We also need to add a generate for binning. Then the silver will recommend binning when the cardinality is too high.

File: draco/asp/define.lp [hunk 3]
```diff
@@ -47,3 +47,9 @@ domain((encoding,stack),(zero;center;normalize)).
 
 % @definition(scale_zero) Scale start with 0.
 domain((scale,zero),true).
+
+% @definition(facets) Definition for facets
+domain((facet,field),F) :- entity(field,_,F).
+domain((facet,channel),(row;col)).
+domain((facet,binning),(10;25;200)).
```
Comment (domoritz) [inline draco/asp/define.lp:54]: Extract the binning options into something we can reuse for the encoding binning.

**Review by domoritz — APPROVED** [review]

File: draco/asp/define.lp (hunk 3 above)
Comment (Zehua-Zeng) [reply draco/asp/define.lp:54]: Do you mean to add a helper to know if a field is binned or not?

Comment (domoritz) [reply draco/asp/define.lp:54]: No, add a fact with `binCounts()` or something so we can reuse the possible bin counts.

**Pull request merged**